crates/oxc_linter/src/plugin.rs

```diff
@@ -0,0 +1,212 @@
+                                    (start[i] as usize, (end[i] - start[i]) as usize),
+                                    input_query.reason.as_str()
+                                )],
+                                "Unexpected error"
+                            ));
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::{env, path::Path, rc::Rc};
+
+    use oxc_allocator::Allocator;
+    use oxc_diagnostics::Report;
+    use oxc_parser::Parser;
+    use oxc_semantic::SemanticBuilder;
+    use oxc_span::SourceType;
+    use path_calculate::Calculate;
+
+    use super::LinterPlugin;
+    use crate::{lint_adapter::SingleTest, Linter};
+
+    #[test]
+    fn test_queries() {
+        env::set_var("OXC_PLUGIN", "./examples/queries");
+        let plugin = LinterPlugin::new();
+        for rule in plugin.rules {
+            for (i, test) in rule.tests.pass.iter().enumerate() {
+                let messages = run_test(test, &rule.name);
+                assert!(
+                    messages.is_empty(),
+                    "{}'s test {} is failing when it should pass.\nErrors: {:#?}\nPath: {}\nCode:\n\n{}\n\n",
+                    rule.name,
+                    i + 1,
+                    messages,
+                    test.file_path,
+                    test.code
+                );
+            }
+
+            for (i, test) in rule.tests.fail.iter().enumerate() {
+                let messages = run_test(test, &rule.name);
+                assert!(
+                    !messages.is_empty(),
+                    "{}'s test {} is passing when it should fail.\nPath: {}\nCode:\n\n{}\n\n",
+                    rule.name,
+                    i + 1,
+                    test.file_path,
+                    test.code
+                );
+            }
+
+            if rule.tests.pass.len() + rule.tests.fail.len() > 0 {
+                println!(
+                    "{} passed {} tests successfully.\n",
+                    rule.name,
+                    rule.tests.pass.len() + rule.tests.fail.len()
+                );
+            }
+        }
+    }
+
+    fn run_test(test: &SingleTest, rule_name: &str) -> Vec<Report> {
+        let file_path = &test.file_path;
+        let source_text = &test.code;
+
+        let path = Path::new(file_path);
+        let allocator = Allocator::default();
+        let source_type = SourceType::from_path(file_path).unwrap();
+        let ret = Parser::new(&allocator, source_text, source_type).parse();
+
+        // Handle parser errors
+        assert!(ret.errors.is_empty(), "Parser errors: {:?} Code:\n\n{:?}", ret.errors, test.code);
+
+        let program = allocator.alloc(ret.program);
+        let semantic_ret = SemanticBuilder::new(source_text, source_type)
+            .with_trivias(&ret.trivias)
+            .build(program);
+
+        let linter = Linter::new().with_fix(false).only_use_query_rule(rule_name);
```

Comment:
Not super important in this pr, but maybe we should also make a plan for fixes within queries too?